proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,754 @@
+invisible quantifiers is necessary to properly handle higher-rank programs. To
+illustrate why, consider the following program ::
+
+  type D :: forall (f :: forall k. @U k -> @U Type) -> @M Type
+  data D f = D (f Bool) (f 0)
+
+  type F :: forall k. @U k -> @U Type
+  type family F a where
+    F 0 = Int
+    F Bool = Char
+
+  p :: D F
+  p = D 'c' 0
+
+Here, ``D`` has a rank-2 kind and its argument is a function. To be able to
+pass in ``F``, the forall must be unmatchable in ``D``'s argument.
+
+We also include invisible non-dependent quantification (``ty =>``), mainly for
+the sake of completeness.
+
+.. _Inference:
+
+Meaning of an annotation-free ``->``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Even though this proposal introduces a way to annotate arrows, in many cases the
+annotations can be inferred. The primary aim of inference is to ease the
+transition as most programs written today can be unambiguously inferred.
+
+The meaning of ``(->)`` depends on the context in which it is written. Below is
+a list of the different contexts with examples.
+
+Data types
+##########
+
+The kind arrows of data types (and data families) are all matchable.
+
+::
+
+  -- inferred:  Type -> @M Type
+  type Maybe :: Type -> Type
+  data Maybe a = ...
+
+here, users are not required to specify ``Type -> @M Type``, as this information
+can be inferred from the data declaration itself.
+
+Higher-order arguments also get assigned matchable
+
+::
+
+  -- inferred: (Type -> @M Type) -> @M Type
+  type HK :: (Type -> Type) -> Type
+  type HK f = ...
+
+Type families
+#############
+
+Type family (and type synonym) *arguments* are unmatchable
+
+::
+
+  -- inferred: Type -> @U Type
+  type Id :: Type -> Type
+  type Id a = a
+
+is unambiguous, and no annotation is required. However, the unambiguity here
+arises not solely of the fact that ``Id`` is a type synonym, but also that it
+binds its argument on the left-hand side. The arrows not corresponding to arguments
+bound on the LHS are inferred to be matchable ::
+
+  -- inferred: Type -> @M Type
+  type MyMaybe :: Type -> Type
+  type family MyMaybe where
+    MyMaybe = Maybe
+
+``MyMaybe`` is a nullary type family, and its return kind is thus matchable (see
+the *Arity of type families* section for more details about type family arities).
+
+Even for type families, higher-order arguments get assigned matchable kinds
+unless specified otherwise
+::
+
+  -- inferred: forall a b. @U (a -> @M b) -> @U [a] -> @U [b]
+  type Map :: (a -> b) -> [a] -> [b]
+  type family Map f xs where ...
+
+Note that the forall is unmatchable, as discussed previously. The function
+argument is matchable, which is consistent with the behaviour today.
+
+Also note that this higher-order defaulting mechanism only applies when a kind
+signature is given. When no signature is given, the inferred kind may be
+matchability-polymorphic ::
+
+  -- inferred: Map :: forall a b m. @U (a -> @m b) -> @U [a] -> @U [b]
```

Comment:
It might be worth briefly justifying the decision not to default matchability variables in the case of a type-level declaration without a standalone kind signature, which differs from how type-level declarations with standalone kind signatures work. For bonus points, compare this to the way defaulting works for levity polymorphism (which is different still).